gobblin-core/src/main/java/gobblin/writer/RetryWriter.java

```diff
@@ -0,0 +1,162 @@
+
+import gobblin.configuration.State;
+import gobblin.instrumented.Instrumented;
+import gobblin.metrics.GobblinMetrics;
+
+import java.io.IOException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.codahale.metrics.Meter;
+import com.github.rholder.retry.Attempt;
+import com.github.rholder.retry.RetryListener;
+import com.github.rholder.retry.Retryer;
+import com.github.rholder.retry.RetryerBuilder;
+import com.github.rholder.retry.StopStrategies;
+import com.github.rholder.retry.WaitStrategies;
+import com.google.common.base.Optional;
+import com.google.common.base.Predicate;
+
+/**
+ * Retry writer follows decorator pattern that retries on inner writer's failure.
+ * @param <D>
+ */
+public class RetryWriter<D> implements DataWriter<D> {
+  private static final Logger LOG = LoggerFactory.getLogger(RetryWriter.class);
+  public static final String FAILED_RETRY_WRITES_METER = "gobblin.writer.failed.retry.writes";
+
+  private final DataWriter<D> writer;
+  private final Retryer<Void> retryer;
+  private boolean isFailure = false;
+
+  public RetryWriter(DataWriter<D> writer, State state) {
+    this.writer = writer;
+    this.retryer = buildRetryer(state);
+  }
+
+  /**
+   * Build Retryer.
+   * - If Writer implements Retriable, it will use the RetryerBuilder from the writer.
+   * - Otherwise, it will use DEFAULT writer builder.
+   *
+   * - If Gobblin metrics is enabled, it will emit all failure count in to metrics.
+   *
+   * @param state
+   * @return
+   */
+  private Retryer<Void> buildRetryer(State state) {
+    RetryerBuilder<Void> builder = null;
+    if (writer instanceof Retriable) {
+      builder = ((Retriable) writer).getRetryerBuilder();
+    } else {
+      builder = getDefaultRetryBuilder();
+    }
+
+    if (GobblinMetrics.isEnabled(state)) {
+      final Optional<Meter> retryMeter = Optional.of(Instrumented.getMetricContext(state, getClass()).meter(FAILED_RETRY_WRITES_METER));
+
+      builder.withRetryListener(new RetryListener() {
+        @Override
+        public <V> void onRetry(Attempt<V> attempt) {
+          if (attempt.hasException()) {
+            LOG.warn("Caught exception. This may be retried.", attempt.getExceptionCause());
+            Instrumented.markMeter(retryMeter);
+          }
+        }
+      });
+    }
+    return builder.build();
+  }
+
+  @Override
+  public void close() throws IOException {
+    writer.close();
+  }
+
+  @Override
+  public void write(final D record) throws IOException {
+    //Need a Callable interface to be wrapped by Retryer.
+    Callable<Void> writeCall = new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        writer.write(record);
+        return null;
+      }
+    };
+
+    try {
+      this.retryer.wrap(writeCall).call();
+    } catch (Exception e) {
+      isFailure = true;
+      throw new RuntimeException(e);
```

Comment:
Since `isFailure` is set and checked in `commit` phase, do you need to throw `RuntimException` here?